csi/controller_server.go

```diff
@@ -335,12 +374,176 @@ func (cs *ControllerServer) GetCapacity(context.Context, *csi.GetCapacityRequest
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-func (cs *ControllerServer) CreateSnapshot(context.Context, *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	logrus.Debugf("ControllerServer CreateSnapshot req: %v", req)
+	csiLabels := req.Parameters
+	csiSnapshotName := req.GetName()
+	csiVolumeName := req.GetSourceVolumeId()
+	if len(csiVolumeName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	} else if len(csiSnapshotName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot name must be provided")
+	}
+
+	// we check for backup existence first, since it's possible
+	// the actual volume is no longer available but the snapshot/backup
+	backupVolume, err := cs.apiClient.BackupVolume.ById(csiVolumeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	backupListOutput, err := cs.apiClient.BackupVolume.ActionBackupList(backupVolume)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// NOTE: csi-snapshots assume a 1 to 1 relationship, longhorn allows for multiple backups of a snapshot
```

Comment:
CSI snapshot are Longhorn backups, so it's still one to one?

Reply:
csi snapshot(1) -> (1)longhorn snapshot(1) -> (n) backups

Reply:
I see. So you're trying to use/match the CSI snapshot name with the snapshot name in Longhorn. So if there is any backup with this snapshot name, then we will use that backup. Otherwise, we will create a new snapshot with the name and back it up.

I didn't notice this part of logic in the LEP, so I didn't realize we're actively using the CSI snapshot name.

Will it make things easier if we just keep the CSI snapshot name in the CSI? So any new snapshot call will create a new snapshot/backup, no matter what the name is?

Reply:
This would not work, the csi-snapshot-request name is generated by the csi snapshotter, based on the configured prefix + `Snapshot.uuid` this information is also used to generate the `SnapshotContent.name = snapcontent-uuid`

example: `kubectl get volumesnaphots test-snapshot-pvc`:
```yaml
kind: VolumeSnapshot
metadata:
  name: test-snapshot-pvc
  uid: aec4bdc3-3244-443e-a26d-4cd48b3e27af
spec:
  source:
    persistentVolumeClaimName: test-vol
  volumeSnapshotClassName: longhorn
status:
  boundVolumeSnapshotContentName: snapcontent-aec4bdc3-3244-443e-a26d-4cd48b3e27af
  readyToUse: false
```

the csi-snapshotter will always call the longhorn csi plugin with the same csi snapshot request name for a specific `Snapshot` object.

So we have to use the csi snapshotter generated name, for the longhorn snapshot or introduce a CRD (VolumeBackup) so that we have a persisted way of associating `csi - snapshot - backup`

If we would ignore the name, we would end up generating a new snapshot for each successive call, since a backup can take a long time for very big volumes, that would not be the desired behavior.

One might consider storing the csi name in memory, but that wouldn't work since there can be multiple csi-plugins also that wouldn't survive crashes of the csi-plugin. Basically the same problem we currently have with the backupStatus information. Long term the only way out of this is the backup refactor tracked here longhorn/longhorn#1761

Reply:
We've discussed this and I am good with the solution.

Also, need to update `snapshotHandle` schema.

Reply:
Updated the schema for the `snapshotID` to use `bs://volumeName/backupName`, inside of the csi driver it's called `snapshotID`, the final VolumeSnapshotContent calls it `snapshotHandle`. Since the csi driver only deals with the request/response pairs I decided to stick with `snapshotID`